### DISCUSSION TITLE
[C++] Merge ATNConfig and LexerATNConfig

### DISCUSSION THREAD

**Discussion by jcking** [antlr/antlr4#3527] [Ideas]
We can potentially remove most shared_ptr\<ATNConfig\> by merging LexerATNConfig into ATNConfig, making it non virtual and concrete. All members would be trivially copyable or shared_ptr themselves. Now we could return ATNConfig by value and avoid all usages of shared_ptr\<ATNConfig\> in the atn code.

@mike-lischke WDYT?

**Comment by KvanTTT** [comment]
I'm not sure it's a good idea. There are other similar entities like `ATNFactory` and `LexerATNFactory`. What's the difference between `ATNConfig` and `LexerATNConfig`?

**Comment by jcking** [comment]
In C++ the difference is basically a single member variable, `std::shared_ptr<LexerActionExecutor>`.

**Comment by mike-lischke** [comment]
The two are used in completely different contexts. One is the lexer, the other is the parser. How would you unite these two classes? Have a single class with all the fields combined from each  former class?

There's more than a single member variable that sets both classes apart. The differences are not huge, but given the sizes of these classes, it doesn't appear to be useful to merge them.

The use of shared_ptr in the prediction code stems from the fact that the configs constitute a graph with no root that controls lifetime. Sometimes configs are created temporarily and replaced. In Java, when you set a new instance of a class in a field, the existing value is automatically destroyed, unless used otherwise. You can use this approach only by using a shared_ptr.

**Comment by jcking** [comment]
ATNConfig is the same as LexerATNConfig except for `std::shared_ptr<LexerActionExecutor>`. If we add `std::shared_ptr<LexerActionExecutor>` to ATNConfig, we can avoid using `std::share_ptr<ATNConfig>` and just pass them around by value and reference. In reality the size of the class will be the same, because we will be removing the vtable pointer as ATNConfig will no longer be virtual. The only things LexerATNConfig overrides is `hashCode()`. It currently doesn't override `operator==` which is probably a bug. So merging them is pretty easy. This leaves us with only having to use `std::shared_ptr` for `SemanticContext` and `PredicateContext`.

**Comment by mike-lischke** [comment]
What about the non-greedy decision check (whose result is also part of the hash)? And can't follow your argumentation of avoiding `std::shared_pts<ATNConfig>` if we remove the dedicated lexer atn config class. As I wrote above, the shared_ptr use is necessary because of the way how the graph is stored. This will not change just because a specialized class is removed.

For the `==` operator overloading, that's something where you probably are right. We need to check the Java source to be sure.
